### PR TITLE
Make new receivers available inside declaration

### PR DESCRIPTION
**Pull request by shadrina**

### CODE REVIEW

File: compiler/resolution/src/org/jetbrains/kotlin/resolve/scopes/utils/ScopeUtils.kt [hunk 1]
```diff
@@ -261,7 +261,7 @@ fun LexicalScope.replaceImportingScopes(importingScopeChain: ImportingScope?): L
 fun LexicalScope.createScopeForDestructuring(newReceiver: ReceiverParameterDescriptor?): LexicalScope {
     return LexicalScopeImpl(
         parent, ownerDescriptor, isOwnerDescriptorAccessibleByLabel,
-        newReceiver,
+        newReceiver?.let { listOf(it) } ?: emptyList(),
```
Comment (dzharkov) [inline compiler/resolution/src/org/jetbrains/kotlin/resolve/scopes/utils/ScopeUtils.kt]: `listOfNotNull()`

Comment (shadrina) [reply compiler/resolution/src/org/jetbrains/kotlin/resolve/scopes/utils/ScopeUtils.kt]: Fixed

File: compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/ClassResolutionScopesSupport.kt [hunk 2]
```diff
@@ -102,7 +102,7 @@ class ClassResolutionScopesSupport(
         val lexicalChainedScope = LexicalChainedScope.create(
             parentForNewScope, ownerDescriptor,
             isOwnerDescriptorAccessibleByLabel = false,
-            implicitReceiver = companionObjectDescriptor?.thisAsReceiverParameter,
+            implicitReceivers = companionObjectDescriptor?.thisAsReceiverParameter?.let { listOf(it) } ?: emptyList(),
```
Comment (dzharkov) [inline compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/ClassResolutionScopesSupport.kt]: `listOfNotNull` too

File: compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/ImplicitScopeTower.kt [hunk 3]
```diff
@@ -86,7 +86,7 @@ class CandidateWithBoundDispatchReceiver(
 )
 
 fun getResultApplicability(diagnostics: Collection<KotlinCallDiagnostic>) =
-    diagnostics.maxBy { it.candidateApplicability }?.candidateApplicability
+    diagnostics.maxByOrNull { it.candidateApplicability }?.candidateApplicability
```
Comment (dzharkov) [inline compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/ImplicitScopeTower.kt]: I didn't get why it's necessary

File: idea/src/org/jetbrains/kotlin/idea/quickfix/replaceWith/ReplaceWithAnnotationAnalyzer.kt [hunk 4]
```diff
@@ -251,7 +251,7 @@ object ReplaceWithAnnotationAnalyzer {
                     propertyHeader,
                     descriptor,
                     false,
-                    descriptor.extensionReceiverParameter,
+                    descriptor.extensionReceiverParameter?.let { listOf(it) } ?: emptyList(),
```
Comment (dzharkov) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/replaceWith/ReplaceWithAnnotationAnalyzer.kt]: `listOfNotNull`

File: compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/TowerResolver.kt [hunk 5]
```diff
@@ -376,7 +376,7 @@ class TowerResolver {
         }
 
         override fun getFinalCandidates(): Collection<C> {
-            val moreSuitableGroup = candidateGroups.minBy { it.groupApplicability } ?: return emptyList()
+            val moreSuitableGroup = candidateGroups.minByOrNull { it.groupApplicability } ?: return emptyList()
```
Comment (dzharkov) [inline compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/TowerResolver.kt:379]: Why did this change happen?

**Review by dzharkov — CHANGES_REQUESTED** [review]

File: compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/ClassResolutionScopesSupport.kt (hunk 2 above)
Comment (shadrina) [reply compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/ClassResolutionScopesSupport.kt]: Done

File: compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/ImplicitScopeTower.kt (hunk 3 above)
Comment (shadrina) [reply compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/ImplicitScopeTower.kt]: Selected `Cleanup` option on commit (`maxBy` is deprecated)

Comment (shadrina) [reply compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/ImplicitScopeTower.kt]: Reverted

File: idea/src/org/jetbrains/kotlin/idea/quickfix/replaceWith/ReplaceWithAnnotationAnalyzer.kt (hunk 4 above)
Comment (shadrina) [reply idea/src/org/jetbrains/kotlin/idea/quickfix/replaceWith/ReplaceWithAnnotationAnalyzer.kt]: Done

File: compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/TowerResolver.kt (hunk 5 above)
Comment (shadrina) [reply compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/TowerResolver.kt:379]: Again `Cleanup` effect. Reverted

**Review by dzharkov — APPROVED** [review]

**Pull request merged**